OPENAPI-04: Add testing capabilities to verify and validate the generated OpenAPI specification (#2)
openmrs/openmrs-openapi-maven-plugin: README.md

# OpenMRS REST OpenAPI Generator

> ![OpenMRS Logo](src/static/openmrs.png)

Reviewer: This will, hopefully, be more than just a GSoC project. Can you remove this image, or replace with something more appropriate?

> A Maven plugin that automatically generates **OpenAPI 3.0 specifications** from OpenMRS REST resources during the build process. This plugin uses runtime introspection to analyze REST resource handlers and produces comprehensive API documentation that's ready for tools like Swagger UI, Postman, or API clients.

Reviewer: ```suggestion
A Maven plugin that, when applied to OpenMRS modules, automatically generates **OpenAPI 3.0 specifications** for the module's REST resources during the build process. This plugin uses runtime introspection to analyze REST resource handlers and produces comprehensive API documentation that's ready for tools like Swagger UI, Postman, or API clients.
```

## 🚀 What This Plugin Does

> The plugin **automatically discovers and analyzes** OpenMRS REST resources in your module, generating:

Reviewer: These are the eventual goals of the plugin, but I think it's important that we:
-  mention that it's a work in progress and not completed yet 
- document what works  (GET routes, the usual representations) and doesn't work (POST / DELETE routes, named representations) as of now. Avoid words like "Complete" until we are actually there.

Maybe also mention that, for now, this plugin has been tested against the REST module and queues module.

- **📋 Complete OpenAPI 3.0 specification** with paths, parameters, and schemas
- **🔍 Resource introspection** using OpenMRS representation system (DEFAULT, FULL, REF)
- **📝 Interactive API documentation** compatible with Swagger UI
- **⚡ Build-time generation** integrated with Maven lifecycle
- **🛡️ Type-safe schemas** with proper validation and examples

Perfect for **API documentation**, **client SDK generation**, **testing**, and **module integration**.

## 🏗️ Architecture

### Forked Process Design
```
Maven Plugin (Lightweight)
    ↓ Spawns
Forked JVM (Full OpenMRS Context)
    ↓ Loads OpenMRS + H2 Database
    ↓ Analyzes REST Resources
    ↓ Generates
OpenAPI 3.0 JSON Specification
```

**Why Forked JVM?** The plugin runs OpenMRS in a separate process to avoid ClassLoader conflicts between Maven and OpenMRS runtime dependencies.

## 📊 Sample Output

> The plugin generates a complete OpenAPI 3.0 specification like this:

Reviewer: remove "complete"

            "schema": { "type": "string" }
          },
          {
            "name": "v",
            "in": "query", 
            "description": "The representation to return. Allowed values: 'default', 'full', 'ref', or custom",
            "schema": {
              "type": "string",
              "enum": ["default", "full", "ref"]
            }
          }
        ],
        "responses": {
          "200": {
            "description": "Successful response",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/QueueDefault"
                }
              }
            }
          }
        }
      }
    }
  },
  "components": {
    "schemas": {
      "QueueDefault": {
        "type": "object",
        "properties": {
          "uuid": { "type": "string" },
          "display": { "type": "string" },
          "name": { "type": "string" },
          "description": { "type": "string" }
        }
      }
    }
  }
}
```

## ⚡ Quick Start

### Prerequisites
- **Java 8+** (OpenMRS requirement)
- **Maven 3.6+**
- **OpenMRS module** with REST resources

### 1. Install the Plugin

> First, build and install the plugin to your local Maven repository:

Reviewer: Is this still accurate? We published the plugin, so no need to install locally right?

```bash
git clone https://github.com/capernix/openmrs-maven-plugin-openapi.git
cd openmrs-maven-plugin-openapi/openmrs-rest-representation-analyzer
mvn clean install
```

### 2. Add to Your OpenMRS Module

Add the plugin to your module's `omod/pom.xml`:

```xml
<build>
  <plugins>
    <!-- OpenMRS REST OpenAPI Generator Plugin -->
    <plugin>

>       <groupId>org.openmrs.plugin</groupId>
>       <artifactId>openmrs-rest-analyzer</artifactId>

Reviewer: ```suggestion
      <groupId>org.openmrs.maven.plugins</groupId>
      <artifactId>openmrs-openapi-maven-plugin</artifactId>
```

            <scanPackages>
              <package>org.openmrs.module.yourmodule.web.resources</package>
            </scanPackages>
          </configuration>
        </execution>
      </executions>
    </plugin>
  </plugins>
</build>
```

### 3. Add Required Test Dependencies

Add these test dependencies to your `omod/pom.xml` (required for OpenMRS context):

```xml
<dependencies>
  <!-- OpenMRS Test Framework -->
  <dependency>
    <groupId>org.openmrs.test</groupId>
    <artifactId>openmrs-test</artifactId>
    <type>pom</type>
    <scope>test</scope>
  </dependency>
  
  <!-- JUnit Platform (for test execution) -->
  <dependency>
    <groupId>org.junit.platform</groupId>
    <artifactId>junit-platform-launcher</artifactId>
    <scope>test</scope>
  </dependency>
  
  <!-- OpenMRS Web Test Support -->
  <dependency>
    <groupId>org.openmrs.web</groupId>
    <artifactId>openmrs-web</artifactId>
    <classifier>tests</classifier>
    <scope>test</scope>
  </dependency>
</dependencies>
```

### 4. Generate OpenAPI Specification

Run the plugin to generate your API spec:

```bash
# Generate during normal build
mvn clean process-classes

# Or run the plugin directly

> mvn openmrs-rest-analyzer:analyze-representations

Reviewer: update

# Full build with spec generation
mvn clean install
```

### 5. View Your API Documentation

The generated OpenAPI specification will be at:
```
target/openapi/openapi.json
```

**Visualize with Swagger UI:**
1. Copy the JSON content
2. Go to https://editor.swagger.io/
3. Paste your specification
4. Enjoy interactive API documentation! 🎉

## 🔧 Usage Commands

### Basic Commands

```bash
# Generate OpenAPI spec during build (recommended)
mvn clean process-classes

# Run plugin directly
mvn openmrs-rest-analyzer:analyze-representations

# Full build with spec generation
mvn clean install

# Debug mode with verbose output
mvn openmrs-rest-analyzer:analyze-representations -X

# Custom timeout for large modules (default: 300 seconds)
mvn openmrs-rest-analyzer:analyze-representations -DtimeoutSeconds=600
```

### Integration with Build Lifecycle

The plugin automatically runs during the `process-classes` phase when configured in your `pom.xml`. No manual execution needed!

```bash
# These commands will automatically generate OpenAPI specs:
mvn compile
mvn test  
mvn package
mvn install
```

## ⚙️ Configuration Options

### Basic Configuration

```xml
<plugin>

>   <groupId>org.openmrs.plugin</groupId>
>   <artifactId>openmrs-rest-analyzer</artifactId>

Reviewer: ```suggestion
  <groupId>org.openmrs.maven.plugins</groupId>
  <artifactId>openmrs-openapi-maven-plugin</artifactId>
```

  <version>1.0.0-SNAPSHOT</version>
  <configuration>
    <!-- Auto-detect resource packages (recommended) -->
    <autoDetectResources>true</autoDetectResources>
    
    <!-- OR specify packages manually -->
    <scanPackages>
      <package>org.openmrs.module.yourmodule.web.resources</package>
      <package>org.openmrs.module.yourmodule.api.resources</package>
    </scanPackages>
    
    <!-- OpenMRS version for compatibility -->
    <openmrsVersion>2.4.x</openmrsVersion>
  </configuration>
</plugin>
```

### Auto-Detection

The plugin automatically detects resource packages based on your module structure:

- `openmrs-module-queue` → `org.openmrs.module.queue.web.resources`
- `queue-omod` (groupId: `org.openmrs.module`) → `org.openmrs.module.queue.web.resources`
- Custom modules follow the same pattern

### Output Location

Generated specifications are always placed in:
```
target/openapi/openapi.json
```

## 🔍 Understanding the Output

### OpenAPI 3.0 Structure

The generated specification includes:

- **🛤️ Paths**: All REST endpoints with HTTP methods
- **📋 Parameters**: Path, query, and header parameters
- **🔄 Responses**: Success and error response schemas

> - **📊 Schemas**: Data models for all representations (DEFAULT, FULL, REF)

Reviewer: remove "all"

        }
      },
      "PatientFull": {
        "type": "object", 
        "properties": {
          "uuid": { "type": "string" },
          "display": { "type": "string" },
          "identifiers": {
            "type": "array",
            "items": { "$ref": "#/components/schemas/PatientIdentifierFull" }
          },
          "person": { "$ref": "#/components/schemas/PersonFull" },
          "voided": { "type": "boolean" },
          "dateCreated": { "type": "string", "format": "date-time" }
        }
      }
    }
  }
}
```

## 🛠️ Development & Testing

### Testing the Plugin

```bash
# Build plugin
cd openmrs-rest-representation-analyzer
mvn clean install

# Test with a sample module
cd ../openmrs-module-queue/omod
mvn clean process-classes

# Check generated spec
cat target/openapi/openapi.json | jq .
```

### Plugin Development Commands

```bash
# Build and test the plugin itself
mvn clean install

# Run plugin tests only
mvn test

# Debug plugin with verbose output
mvn clean install -X

# Analyze plugin dependencies
mvn dependency:tree
```

## 🚀 Integration Examples

### Example 1: Queue Module

```xml
<!-- In openmrs-module-queue/omod/pom.xml -->
<plugin>

>   <groupId>org.openmrs.plugin</groupId>
>   <artifactId>openmrs-rest-analyzer</artifactId>

Reviewer: ```suggestion
  <groupId>org.openmrs.maven.plugins</groupId>
  <artifactId>openmrs-openapi-maven-plugin</artifactId>
```

          <package>org.openmrs.module.queue.web.resources</package>
        </scanPackages>
      </configuration>
    </execution>
  </executions>
</plugin>
```

**Generated endpoints:**
- `GET /ws/rest/v1/queue/{uuid}`
- `POST /ws/rest/v1/queue`
- `GET /ws/rest/v1/queue-room/{uuid}`
- `GET /ws/rest/v1/queue-entry/{uuid}`

### Example 2: Webservices REST Module

```xml
<!-- In openmrs-module-webservices.rest/omod/pom.xml -->
<plugin>
  <groupId>org.openmrs.plugin</groupId>
  <artifactId>openmrs-rest-analyzer</artifactId>
  <version>1.0.0-SNAPSHOT</version>
  <executions>
    <execution>
      <id>generate-openapi-spec</id>
      <phase>process-classes</phase>
      <goals>
        <goal>analyze-representations</goal>
      </goals>
      <configuration>
        <autoDetectResources>true</autoDetectResources>
      </configuration>
    </execution>
  </executions>
</plugin>
```

**Generated endpoints:** 100+ core OpenMRS REST endpoints including:
- Patient, Person, Encounter, Observation resources
- Concept, ConceptClass, ConceptDatatype resources  
- User, Role, Privilege resources

### Example 3: Custom Module

```xml
<!-- For any custom OpenMRS module -->
<plugin>

>   <groupId>org.openmrs.plugin</groupId>
>   <artifactId>openmrs-rest-analyzer</artifactId>

Reviewer: ```suggestion
  <groupId>org.openmrs.maven.plugins</groupId>
  <artifactId>openmrs-openapi-maven-plugin</artifactId>
```

  <version>1.0.0-SNAPSHOT</version>
  <executions>
    <execution>
      <phase>process-classes</phase>
      <goals>
        <goal>analyze-representations</goal>
      </goals>
      <configuration>
        <scanPackages>
          <package>org.openmrs.module.mymodule.web.resources</package>
          <package>org.openmrs.module.mymodule.rest</package>
        </scanPackages>
      </configuration>
    </execution>
  </executions>
</plugin>
```

## 🔧 Troubleshooting

### Common Issues

**1. ClassNotFoundException during execution**
```bash
# Solution: Ensure test dependencies are present
<dependency>
  <groupId>org.openmrs.test</groupId>
  <artifactId>openmrs-test</artifactId>
  <type>pom</type>
  <scope>test</scope>
</dependency>
```

**2. Empty OpenAPI specification**
```bash
# Solution: Check resource package paths

> mvn openmrs-rest-analyzer:analyze-representations -X

Reviewer: Not sure how this works, but I think we need to update the mvn commands in this section with the corresponding name of the plugin `openmrs-openapi-maven-plugin`

cd openmrs-rest-representation-analyzer
mvn clean install
```

### Debug Output

Enable debug logging to see what the plugin is doing:

```bash
mvn openmrs-rest-analyzer:analyze-representations -X
```

This shows:
- Classpath construction
- Resource discovery process
- OpenMRS context initialization
- Schema generation details

## 📦 Dependencies

### Auto-Resolved Dependencies
The plugin automatically handles these dependencies:

- **OpenMRS API** 2.4+ (Core platform)
- **OpenMRS Web Services REST** 2.50+ (REST framework)
- **Jackson** 2.11+ (JSON processing)
- **JUnit Platform** (Test execution)
- **H2 Database** (In-memory testing)
- **Spring Framework** (Dependency injection)

### Required Module Dependencies
Your OpenMRS module must include:

```xml
<dependencies>
  <!-- OpenMRS Web Services REST -->
  <dependency>
    <groupId>org.openmrs.module</groupId>
    <artifactId>webservices.rest-omod</artifactId>
  </dependency>
  
  <!-- Test Framework (required for plugin) -->
  <dependency>
    <groupId>org.openmrs.test</groupId>
    <artifactId>openmrs-test</artifactId>
    <type>pom</type>
    <scope>test</scope>
  </dependency>
</dependencies>
```

> ## 🚀 Using the Generated Specification

Reviewer: I don't think we need this section. It'll be incomplete at best. If you want to, just mention in a couple sentences the different ways you can use the OpenAPI JSON.
